app/src/main/java/com/github/se/orator/model/profile/UserProfileRepositoryFirestore.kt

```diff
@@ -340,6 +350,372 @@ class UserProfileRepositoryFirestore(private val db: FirebaseFirestore) : UserPr
+          onFailure(exception)
+        }
+  }
+
+  /**
+   * Get profiles the user received requests from based on the UIDs stored in the user's profile.
+   *
+   * @param recReqUIds List of UIDs of the users the user received requests from to be retrieved.
+   * @param onSuccess Callback to be invoked with the list of friends' profiles.
+   * @param onFailure Callback to be invoked on failure with the exception.
+   */
+  override fun getSentReqProfiles(
+      friendUids: List<String>,
+      onSuccess: (List<UserProfile>) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    if (friendUids.isEmpty()) {
+      onSuccess(emptyList()) // Return an empty list if no friends
+      return
+    }
+
+    db.collection(collectionPath)
+        .whereIn("uid", friendUids)
+        .get()
+        .addOnSuccessListener { querySnapshot ->
+          val friends = querySnapshot.documents.mapNotNull { documentToUserProfile(it) }
+          onSuccess(friends)
+        }
+        .addOnFailureListener { exception ->
+          Log.e("UserProfileRepository", "Error fetching friends profiles", exception)
+          onFailure(exception)
+        }
+  }
+
+  /**
+   * Sends a friend request from the current user to another user.
+   *
+   * @param currentUid The UID of the user sending the request.
+   * @param friendUid The UID of the user receiving the request.
+   * @param onSuccess Callback invoked on successful operation.
+   * @param onFailure Callback invoked with an [Exception] on failure.
+   */
+  override fun sendFriendRequest(
+      currentUid: String,
+      friendUid: String,
+      onSuccess: () -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    // Reference to the current user's document
+    val currentUserRef = db.collection(collectionPath).document(currentUid)
+    // Reference to the friend user's document
+    val friendUserRef = db.collection(collectionPath).document(friendUid)
+
+    // Run a Firestore transaction to ensure atomicity
+    db.runTransaction { transaction ->
+          // Get the current user's snapshot
+          val currentUserSnapshot = transaction.get(currentUserRef)
+          // Get the friend user's snapshot
+          val friendUserSnapshot = transaction.get(friendUserRef)
+
+          // Retrieve the current user's sent requests
+          val currentSentReq = currentUserSnapshot.get("sentReq") as? List<String> ?: emptyList()
+          // Retrieve the friend user's received requests
+          val friendRecReq = friendUserSnapshot.get("recReq") as? List<String> ?: emptyList()
+
+          // Check if the friend request has already been sent
+          if (currentSentReq.contains(friendUid)) {
+            throw Exception("Friend request already sent.")
+          }
+
+          // Optionally, check if the friend has already sent a request to the current user
+          if (friendRecReq.contains(currentUid)) {
+            throw Exception("The user has already sent you a friend request.")
```

Comment:
Maybe it would be more logical to then accept the person's request instead of just failing. But this is just my point of view, not an issue.

Reply:
This will never be the case as I am handling it with the AlertDialog, so we will never have a case where we call sendFriendRequest and the user has already received a friend request from the same profile. I removed this condition.